macrosynergy/signal/target_positions.py

```diff
@@ -104,9 +106,102 @@ def cs_unit_returns(df: pd.DataFrame, contract_returns: List[str],
+    """
+
+    split = lambda b: b.split('_')[0]
+
+    cross_sections = list(map(split, contracts))
+    basket = Basket(df=df, contracts=contracts, ret=ret, cry=None, start=start, end=end,
+                    blacklist=None, wgt=None)
+
+    dfw_wgs = basket.weight_dateframe(weight_meth="equal", weights=None, max_weight=1.0,
+                                      remove_zeros=True)
+    dfw_wgs = dfw_wgs.reindex(sorted(dfw_wgs.columns), axis=1)
+
+    # Reduce to the cross-sections held in the respective basket.
+    df_mods_w = df_mods_w[cross_sections]
+    df_mods_w = df_mods_w.reindex(sorted(df_mods_w.columns), axis=1)
+
+    # Adjust the target positions to reflect the weighting method.
+    df_mods_w = df_mods_w.multiply(dfw_wgs.to_numpy())
+
+    return df_mods_w
+
+def consolidation_help(panel_df: pd.DataFrame, basket_df: pd.DataFrame):
+    """
+    The function receives a panel dataframe and a basket of cross-sections of the same
+    contract type. Therefore, aim to consolidate the targeted positions across the shared
+    contracts.
+
+    :param <pd.DataFrame> panel_df:
+    :param <pd.DataFrame> basket_df:
+
+    :return <pd.DataFrame, pd.DataFrame> returns the consolidated and reduced dataframes.
+    """
+
+    basket_cids = basket_df['cid'].unique()
+    panel_cids = panel_df['cid'].unique()
+
+    basket_df_c = basket_df.copy()
+    for cid in basket_cids:
+        if cid in panel_cids:
+            basket_indices = basket_df['cid'] == cid
+            df_1 = basket_df[basket_indices]
+            indices = panel_df['cid'] == cid
+            panel_df[indices]['value'] += df_1['value']
+            basket_indices = ~basket_indices
+            basket_df = basket_df[basket_indices]
+        else:
+            continue
+
+    return panel_df, basket_df.reset_index()
+
+def consolidation_driver(data_frames: List[pd.DataFrame], ctypes: List[str]):
+    """
+    Method used to consolidate positions if baskets are used. The constituents of a
+    basket will be a subset of one of the panels.
+
+    :param <List[pd.DataFrame]> data_frames: list of the target position dataframes.
+    :param <List[str]> ctypes:
+
+    :return <List[pd.DataFrame]> list of dataframes having consolidated positions.
+    """
+
+    no_ctypes = len(ctypes)
+    dict_ = dict(zip(ctypes[:no_ctypes], data_frames[:no_ctypes]))
+    df_baskets = data_frames[no_ctypes:]
+
+    split_2 = lambda b: b.split('_')[1]
+    # Iterating exclusively through the basket dataframes.
+    reduced_baskets = []
+    for df in df_baskets:
+        category = list(map(split_2, df['xcat'].to_numpy()))
+        c_type = category[0]
+
+        panel_df = dict_[c_type]
+        panel_df, basket_df = consolidation_help(panel_df, basket_df=df)
+        dict_[c_type] = panel_df
+        reduced_baskets.append(basket_df)
+
+    return list(dict_.values()) + reduced_baskets
 
 def target_positions(df: pd.DataFrame, cids: List[str], xcat_sig: str, ctypes: List[str],
```

Comment:
The target position function needs to use a previously created weight dataframe (from the basket class) as argument, possibly through the `baskets` argument. For example {'GLB_FX': df_gfx_wgts} would define name and contract weights of a global FX basket.  Importantly, this requires the weight df created in the basket class to contain the contract base tickers as column names.